src/parser/ast.rs

```diff
@@ -517,6 +724,99 @@ impl Neg for Expr {
     }
 }
 
+/// check_ast checks the validity of the provided AST. This includes type checking.
+/// Recursively check correct typing for child nodes and raise errors in case of bad typing.
+pub fn check_ast(expr: Expr) -> Result<Expr, String> {
+    match expr {
+        Expr::Binary(ex) => check_ast_for_binary_expr(ex),
+        // TODO: check other exprs
+        _ => Ok(expr),
+    }
+}
+
+/// the original logic is redundant in prometheus, and the following coding blocks
+/// have been optimized for readability, but all logic SHOULD be covered.
+fn check_ast_for_binary_expr(mut ex: BinaryExpr) -> Result<Expr, String> {
+    if !ex.op.is_operator() {
+        let val = ex.op.val;
+        return Err(format!(
+            "binary expression does not support operator '{val}'"
+        ));
+    }
+
+    if ex.return_bool() && !ex.op.is_comparison_operator() {
+        return Err("bool modifier can only be used on comparison operators".into());
+    }
+
+    if ex.op.is_comparison_operator()
+        && ex.lhs.value_type() == ValueType::Scalar
+        && ex.rhs.value_type() == ValueType::Scalar
+        && !ex.return_bool()
+    {
+        return Err("comparisons between scalars must use BOOL modifier".into());
+    }
+
+    // For `on` matching, a label can only appear in one of the lists.
+    // Every time series of the result vector must be uniquely identifiable.
+    if ex.is_matching_on() && ex.is_labels_joint() {
+        if let Some(labels) = ex.intersect_labels() {
+            if !labels.is_empty() {
+                let label = labels[0];
```

Comment:
Another approach is using `first()`
```suggestion
            if let Some(label) = labels.first() {
```